hypervisor/esxi/esxi.go

```diff
@@ -0,0 +1,309 @@
+func init() {
+	hypervisor.RegisterProvider("esxi", &EsxiHypervisorProvider{})
+	viper.SetDefault("hypervisor.esxi-insecure", true)
+}
+
+func (p *EsxiHypervisorProvider) LoadConfig(sub *viper.Viper) error {
+	if sub.IsSet("bridges.name") {
+		settings.BridgeName = sub.GetString("bridges.name")
+		if sub.IsSet("bridges.type") {
+			switch sub.GetString("bridges.type") {
+			case "linux":
+				settings.BridgeType = Linux
+			case "ovs":
+				settings.BridgeType = OVS
+			default:
+				return errors.Errorf("Unknown bridges.type value: %s", sub.GetString("bridges.type"))
+			}
+		}
+	} else if sub.IsSet("bridges.linux.name") {
+		log.Warn("bridges.linux.name is obsolete option")
+		settings.BridgeName = sub.GetString("bridges.linux.name")
+		settings.BridgeType = Linux
+	} else if sub.IsSet("bridges.ovs.name") {
+		log.Warn("bridges.ovs.name is obsolete option")
+		settings.BridgeName = sub.GetString("bridges.ovs.name")
+		settings.BridgeType = OVS
+	}
+
+	settings.ScriptPath = sub.GetString("hypervisor.script-path")
+	settings.EsxiUser = sub.GetString("hypervisor.esxi-user")
+	settings.EsxiPass = sub.GetString("hypervisor.esxi-pass")
+	settings.EsxiIp = sub.GetString("hypervisor.esxi-ip")
+	settings.EsxiDatacenter = sub.GetString("hypervisor.esxi-datacenter")
+	settings.EsxiInsecure = sub.GetBool("hypervisor.esxi-insecure")
+	settings.EsxiHostSshkey = sub.GetString("hypervisor.esxi-host-sshkey")
+	settings.EsxiVmUser = sub.GetString("hypervisor.esxi-vm-user")
+	settings.EsxiVmPass = sub.GetString("hypervisor.esxi-vm-pass")
+	settings.EsxiVmDatastore = sub.GetString("hypervisor.esxi-vm-datastore")
+
+	esxiInfo := fmt.Sprintf("%s:%s@%s", settings.EsxiUser, settings.EsxiPass, settings.EsxiIp)
+	u, _ := url.Parse("https://")
+	u.Path = path.Join(u.Path, esxiInfo, "sdk")
+	settings.EsxiUrl = u.String()
+
+	return nil
+}
+
+func (p *EsxiHypervisorProvider) CreateDriver(instance *model.Instance, template model.ResourceTemplate) (hypervisor.HypervisorDriver, error) {
+	EsxiTmpl, ok := template.(*model.EsxiTemplate)
+	if !ok {
+		return nil, errors.Errorf("template type is not *model.WmwareTemplate: %T, template")
+	}
+	instanceIdx, _ := strconv.Atoi(strings.TrimPrefix(instance.GetId(), "i-"))
+	driver := &EsxiHypervisorDriver{
+		Base: hypervisor.Base{
+			Log:      log.WithFields(log.Fields{"Hypervisor": "esxi", "instance_id": instance.GetId()}),
+			Instance: instance,
+		},
+		template: EsxiTmpl,
+		machine:  &EsxiMachine{SerialConsolePort: (15000 + instanceIdx)},
+		vmName:   instance.GetId(),
+	}
+
+	return driver, nil
+}
+
+func (d *EsxiHypervisorDriver) log() *log.Entry {
+	return d.Base.Log
+}
+
+func esxiCmd(args ...string) {
+	var a []string
+
+	a = append(a, args[0])
+	a = append(a, fmt.Sprintf("-dc=%s", settings.EsxiDatacenter))
+	a = append(a, fmt.Sprintf("-k=%s", strconv.FormatBool(settings.EsxiInsecure)))
+	a = append(a, fmt.Sprintf("-u=%s", settings.EsxiUrl))
+
+	for i := 1; i < len(args); i++ {
+		a = append(a, args[i])
+	}
+
+	cli.Run(a)
```

Comment:
The `cli.Run()` function returns an int, `0` on success, `1` on failure. The error message is printed directly to os.Stderr by the function itself. 

The options here seems to be to either return an error from the `esxiCmd()` if the return code was `1`, 
or return the function call itself and let the calling function return the error based on the return code